### PR TITLE
feat: Rebalancing votes

### LINKED ISSUE #29: feat: guard algo from overdilution when voting for pools
Take the median bribe price.
And then don’t vote to claim more than 80% of that on any pool

At the median - 57 cents a vote.
If there’s 57k
That’s 100k votes.
Don’t vote with more than 80

### PR DESCRIPTION
**Pull request by SHAKOTN**
Fixes #29

### CODE REVIEW

File: voter/vote.py [hunk 1]
```diff
@@ -21,6 +21,7 @@ def analyze_and_vote():
     treasury_bvecvx = get_treasury_controlled_bvecvx_amount()
     console.print(f"Treasury owned bveCVX amount is: {treasury_bvecvx} \n")
     snapshot = get_gauge_weight_snapshot()
+    # snapshot = get_snapshot_by_id(TEST_SNAPSHOT_ID)
```
Comment (mitche50) [inline voter/vote.py:24]: nit: remove comment

**Review by mitche50 — APPROVED** [review]
Beautiful work Shak!

File: voter/vote.py (hunk 1 above)
Comment (SHAKOTN) [reply voter/vote.py]: Done

File: voter/vote.py [hunk 2]
```diff
@@ -35,21 +35,39 @@ def analyze_and_vote():
...
             ]
         )
     console.print(
         tabulate(suggesting_votes_table,
                  tablefmt="orgtbl",
                  headers=[
-                     "Pool name", "Suggested weighed vote", "vlCVX value for weighting vote",
-                     "$/vlCVX",
+                     "Pool name", "Suggested weighed vote", "CVX To Vote",
+                     "$/vlCVX", "Projected $/vlCVX after Badger vote", "Badger projected output",
```
Comment (btcookies) [inline voter/vote.py:66]: maybe change `Badger projected output` to `Badger projected $ collected`

File: voter/voting_algorithms/proportional_voting.py [hunk 3]
```diff
@@ -29,10 +32,12 @@ class ProportionalVoter:
         # Settings for proportions in voting
         proportions={
             # 5% always votes for f-badgerwbtc
-            'f-badgerwbtc': Decimal(5)
+            'f-badgerwbtc': Decimal(5)  # in %
         },
-        minimal_dollar_vlcvx=Decimal(0.30),
-        parts=5,
+        minimal_dollar_vlcvx=Decimal(0.10),
```
Comment (btcookies) [inline voter/voting_algorithms/proportional_voting.py:37]: in the future i think this should change to minimal_dollar_bribes_collected, we don't care what $/vlcvx is really we just don't want to process selling < $5k or so tokens

File: voter/voting_algorithms/proportional_voting.py [hunk 4]
```diff
@@ -42,7 +47,9 @@ def __init__(
         # Filter out pools that have blacklisted tokens and low values $/vlCVX
         dollar_vlcvx = {
             k: v for k, v in dollar_vlcvx.items()
-            if v['token'] not in self.ALGORITHM_SETTINGS.excluded_tokens
+            # Filter out pools even if they have a single token in rewards that we do not
+            # support
+            if not any(item in v['tokens'] for item in self.ALGORITHM_SETTINGS.excluded_tokens)
```
Comment (btcookies) [inline voter/voting_algorithms/proportional_voting.py:52]: for readability maybe consider separating this filter check into a function like `is_bribe_tokens_valid` cause this breaks brain lol and can get rid of comment

File: voter/voting_algorithms/proportional_voting.py [hunk 5]
```diff
@@ -29,10 +32,12 @@ class ProportionalVoter:
...
             # 5% always votes for f-badgerwbtc
-            'f-badgerwbtc': Decimal(5)
+            'f-badgerwbtc': Decimal(5)  # in %
         },
-        minimal_dollar_vlcvx=Decimal(0.30),
-        parts=5,
+        minimal_dollar_vlcvx=Decimal(0.10),
+        rebalancing_iterations=10000,
+        rebalancing_percent=Decimal(0.1),  # in %
+        min_pool_weight_threshold=Decimal(0.5)  # in %
```
Comment (btcookies) [inline voter/voting_algorithms/proportional_voting.py:40]: what happens if this is 0? would think it would be most optimal?

Comment (SHAKOTN) [reply voter/voting_algorithms/proportional_voting.py:40]: This is the subject to change of course, but can be that for now it's good to have at least 0.5% on some pool

File: voter/voting_algorithms/proportional_voting.py [hunk 6]
```diff
@@ -60,41 +67,86 @@ def __init__(
...
+                    rebalanced_votes[pool] - self.ALGORITHM_SETTINGS.rebalancing_percent
+                # Don't allow for pool weight to drop below threshold to avoid negative numbers
+                if rebal_pool_vote_weight <= self.ALGORITHM_SETTINGS.min_pool_weight_threshold:
+                    continue
+                # Actual rebalancing happening here
+                # Remove pool
+                rebalanced_votes.pop(pool)
+                for rebal_pool, weight in rebalanced_votes.items():
+                    # Check that we add percent only to the pool that has max $/vlCVX
+                    if potential_dollar_vlcvx[rebal_pool] == max(potential_dollar_vlcvx.values()):
```
Comment (btcookies) [inline voter/voting_algorithms/proportional_voting.py:119]: how many iterations is it actually improving the overall $ bribes we receive? curious if it reaches optimization after n-m tries and there are m iterations where it's adding percent to the same pool. could be improved maybe?

Comment (SHAKOTN) [reply voter/voting_algorithms/proportional_voting.py:119]: That is true, I've noticed that after some iteration data is already balanced. But it's hard to identify this state, so I leave it running as it is now

File: voter/voting_algorithms/proportional_voting.py (hunk 3 above)
Comment (SHAKOTN) [reply voter/voting_algorithms/proportional_voting.py:37]: Yeah, let's see if this is an issue and if so, I can add a check against that

File: voter/vote.py (hunk 2 above)
Comment (SHAKOTN) [reply voter/vote.py]: Dine

File: voter/voting_algorithms/proportional_voting.py (hunk 4 above)
Comment (SHAKOTN) [reply voter/voting_algorithms/proportional_voting.py]: Done

**Pull request merged**